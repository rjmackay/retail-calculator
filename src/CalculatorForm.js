import React, { Component } from 'react';
import './CalculatorForm.css';

export const actions = {
  init() {
    return {
      numberOfItems: '',
      pricePerItem: '',
      stateCode: ''
    };
  },
  setNumberOfItems(numberOfItems) {
    return { numberOfItems };
  },
  setPricePerItem(pricePerItem) {
    return { pricePerItem };
  },
  setStateCode(stateCode) {
    return { stateCode };
  }
};

class CalculatorForm extends Component {
  constructor(props) {
    super(props);
    this.state = actions.init();

    this.handleNumberOfItemsChange = this.handleNumberOfItemsChange.bind(this);
    this.handlePricePerItemChange = this.handlePricePerItemChange.bind(this);
    this.handleStateCodeChange = this.handleStateCodeChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleNumberOfItemsChange(event) {
    this.setState(actions.setNumberOfItems(event.target.value));
  }

  handlePricePerItemChange(event) {
    this.setState(actions.setPricePerItem(event.target.value));
  }

  handleStateCodeChange(event) {
    this.setState(actions.setStateCode(event.target.value));
  }

  handleSubmit(event) {
    event.preventDefault();

    this.props.onSubmit(this.state);
  }

  render() {
    return (
      <div className="CalculatorForm">
        <h2>Purchase Form</h2>
        <form onSubmit={this.handleSubmit}>
          <div>
            <label>
              Number of items:
              <input type="text" name="numberOfItems" value={this.state.numberOfItems} onChange={this.handleNumberOfItemsChange} />
            </label>
          </div>
          <div>
            <label>
              Price per item:
              <input type="text" name="pricePerItem" value={this.state.pricePerItem} onChange={this.handlePricePerItemChange} />
            </label>
          </div>
          <div>
            <label>
              State code:
              <input type="text" name="stateCode" value={this.state.stateCode} onChange={this.handleStateCodeChange} />
            </label>
          </div>
          <input type="submit" value="Submit"/>
        </form>
      </div>
    );
  }
}

export default CalculatorForm;
