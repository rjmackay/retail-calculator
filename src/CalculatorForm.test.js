import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import CalculatorForm, { actions } from './CalculatorForm';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<CalculatorForm />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('saves numberOfItems to state and outputs it', () => {
  const component = renderer.create(
    <CalculatorForm />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger state changes
  // Set number of items
  let itemInput = component.root.findByProps({name : 'numberOfItems'});
  itemInput.props.onChange({ target: { value: '100' } });

  // Set price per item
  let priceInput = component.root.findByProps({name : 'pricePerItem'});
  priceInput.props.onChange({ target: { value: '200' } });

  // Submit the form
  let submit = component.root.findByType('form');
  submit.props.onSubmit({
    preventDefault : () => {}
  });

  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('actions', () => {
  it('setNumberOfItems', () => {
    expect(actions.setNumberOfItems(100)
    ).toEqual({
      numberOfItems: 100
    })
  });
  it('setPricePerItem', () => {
    expect(actions.setPricePerItem(100)
    ).toEqual({
      pricePerItem: 100
    })
  });
  it('saveLastSubmission', () => {
    expect(actions.saveLastSubmission({
        numberOfItems: 100,
        pricePerItem: 200,
        lastSubmission: {}
      })
    ).toEqual({
      numberOfItems: 100,
      pricePerItem: 200,
      lastSubmission: {
        numberOfItems: 100,
        pricePerItem: 200,
      }
    })

    expect(actions.saveLastSubmission({
        numberOfItems: 1000,
        pricePerItem: 3000,
        lastSubmission: {
          numberOfItems: 100,
          pricePerItem: 200,
        }
      })
    ).toEqual({
      numberOfItems: 1000,
      pricePerItem: 3000,
      lastSubmission: {
        numberOfItems: 1000,
        pricePerItem: 3000,
      }
    })
  });
})