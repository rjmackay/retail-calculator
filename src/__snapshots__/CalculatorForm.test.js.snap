// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has a numberOfItemsInput 1`] = `
<div
  className="CalculatorForm"
>
  <h2>
    Purchase Form
  </h2>
  <form
    onSubmit={[Function]}
  >
    <label>
      Number of items:
      <input
        name="numberOfItems"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <h3>
    Results:
  </h3>
  <p>
    Number of items: 
  </p>
</div>
`;

exports[`has a numberOfItemsInput 2`] = `
<div
  className="CalculatorForm"
>
  <h2>
    Purchase Form
  </h2>
  <form
    onSubmit={[Function]}
  >
    <label>
      Number of items:
      <input
        name="numberOfItems"
        onChange={[Function]}
        type="text"
        value="100"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <h3>
    Results:
  </h3>
  <p>
    Number of items: 
    100
  </p>
</div>
`;

exports[`saves numberOfItems to state and outputs it 1`] = `
<div
  className="CalculatorForm"
>
  <h2>
    Purchase Form
  </h2>
  <form
    onSubmit={[Function]}
  >
    <label>
      Number of items:
      <input
        name="numberOfItems"
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <h3>
    Results:
  </h3>
  <p>
    Number of items: 
  </p>
</div>
`;

exports[`saves numberOfItems to state and outputs it 2`] = `
<div
  className="CalculatorForm"
>
  <h2>
    Purchase Form
  </h2>
  <form
    onSubmit={[Function]}
  >
    <label>
      Number of items:
      <input
        name="numberOfItems"
        onChange={[Function]}
        type="text"
        value="100"
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <h3>
    Results:
  </h3>
  <p>
    Number of items: 
    100
  </p>
</div>
`;
